### PR TITLE
[MLIR][LLVM]: Add an IR utility to perform slice walking

### PR DESCRIPTION
**Pull request by Dinistro**
This commit introduces a slicing utility that can be used to walk arbitrary IR slices. It additionally ships logic to determine control flow predecessors, which allows users to walk backward slices without dealing with both `RegionBranchOpInterface` and `BranchOpInterface`. 

This utility is used to improve the `noalias` propagation in the LLVM dialect's inliner interface. Before this change, it broke down as soon as pointer were passed through region control flow operations.

### CODE REVIEW

File: mlir/include/mlir/IR/SliceSupport.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+/// backward slice. The callback takes the current value, and returns the walk
+/// continuation, which determines if the walk should proceed and if yes, with
+/// which values.
+using WalkCallback = mlir::function_ref<WalkContinuation(mlir::Value)>;
+
+/// Walks the backward slice starting from the `rootValues` using a depth-first
+/// traversal following the use-def chains. The walk calls the provided
+/// `walkCallback` for each value encountered in the backward slice and uses the
+/// returned walk continuation to determine how to proceed. Additionally, the
+/// walk also transparently traverses through select operations and control flow
```
Comment (gysit) [inline mlir/include/mlir/IR/SliceSupport.h:116]:
```suggestion
/// walk also transparently traverses through control flow
```
nit: I believe we dropped this?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp [hunk 2]
```diff
@@ -221,86 +222,28 @@ static ArrayAttr concatArrayAttr(ArrayAttr lhs, ArrayAttr rhs) {
...
-      if (!anyUnknown)
-        llvm::append_range(workList, operands);
+    // TODO: Add a SelectLikeOpInterface and use it in the slicing utility.
+    if (auto selectOp = val.getDefiningOp<LLVM::SelectOp>())
+      return WalkContinuation::advanceTo(
+          {selectOp.getTrueValue(), selectOp.getFalseValue()});
 
-      continue;
-    }
+    if (isa<OpResult>(val))
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp:242]: At least function arguments should probably be added as well? The cleaner solution maybe to check if the block arguments do not belong to a RegionBranchOp or are not have a BranchOpInterface predecessor add them to the result array and return skip?

**Review by gysit — COMMENTED** [review]
Nice! 

Added two first comments / questions.

**Comment by River707** [comment]
I dont understand party of the description. Why is this in the IR library? The reasons for putting something in IR/ for mlir seem completely unrelated to llvm.

**Review by River707 — DISMISSED** [review]
Marked changes.

**Comment by Dinistro** [comment]
> I dont understand part of the description. Why is this in the IR library? The reasons for putting something in IR/ for mlir seem completely unrelated to llvm.

Sorry, I was referring to the LLVM dialect's inliner interface. It is part of LLVMDialect due to the dialect initializing all interfaces directly in its initialization function.

**Comment by River707** [comment]
> > I dont understand part of the description. Why is this in the IR library? The reasons for putting something in IR/ for mlir seem completely unrelated to llvm.
> 
> Sorry, I was referring to the LLVM dialect's inliner interface. It is part of MLIRLLVMDialect due to the dialect initializing all interfaces directly in its initialization function. As far as I can tell, a dialect library is not supposed to depend on MLIRAnalysis, so that's why I put the utility into MLIRIR.

Ah, I see, thanks for clarifying. I don't think that's a good case for putting something in IR/ though. I'd rather figure out the organizational structure outside of it, and keep the bar for growing IR (which impacts everything) high.

**Comment by Dinistro** [comment]
I made the utility a more generic slice walker and moved the control flow predecessor logic into a separate function. This change allows the usage of the same utility for more arbitrary walks over values, while still providing the necessary logic to perform a backwards slice walk.

I'll now start investigating what is necessary to move this from the IR library into some other place.

File: mlir/include/mlir/IR/SliceSupport.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      : action(WalkAction::Interrupt) {}
+
+  /// Creates a continuation that interrupts the walk.
+  static WalkContinuation interrupt() {
+    return WalkContinuation(WalkAction::Interrupt, {});
+  }
+
+  /// Creates a continuation that adds the user-specified `nextValues` to the
+  /// work list and advances the walk. Unlike advance, this function does not
+  /// add the control flow predecessor values to the work list.
```
Comment (gysit) [inline mlir/include/mlir/IR/SliceSupport.h:49]:
```suggestion
  /// work list and advances the walk.
```
nit: this is not needed anymore.

File: mlir/include/mlir/IR/SliceSupport.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+/// Walks the slice starting from the `rootValues` using a depth-first
+/// traversal. The walk calls the provided `walkCallback` for each value
+/// encountered in the slice and uses the returned walk continuation to
+/// determine how to proceed.
+WalkContinuation walkSlice(mlir::ValueRange rootValues,
+                           WalkCallback walkCallback);
+
+/// Computes a vector of all control predecessors of `value`. Relies on
+/// RegionBranchOpInterface and BranchOpInterface to determine predecessors.
+/// Returns nullopt if value has no predecessors or when the relevant operations
```
Comment (gysit) [inline mlir/include/mlir/IR/SliceSupport.h:93]:
```suggestion
/// Returns nullopt if `value` has no predecessors or when the relevant operations
```
nit: alternative if the value has...

File: mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp [hunk 5]
```diff
@@ -221,86 +222,45 @@ static ArrayAttr concatArrayAttr(ArrayAttr lhs, ArrayAttr rhs) {
...
-      }
-
-      if (!anyUnknown)
-        llvm::append_range(workList, operands);
-
-      continue;
-    }
+    // If this place is reached, `val` is a block argument that is not
+    // understood. Therfore, we conservatively interrupt.
+    // Note: Dealing with the function arguments is not necessary, as the slice
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp:257]:
```suggestion
    // understood. Therefore, we conservatively interrupt.
    // Note: Dealing with function arguments is not necessary, as the slice
```

File: mlir/lib/IR/SliceSupport.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        predecessorOperands.push_back(operands[operandNumber]);
+      }
+    }
+  }
+
+  return predecessorOperands;
+}
+
+/// Returns the predecessor branch operands that match `blockArg`. Returns a
+/// nullopt when some of the predecessor terminators do not implement the
```
Comment (gysit) [inline mlir/lib/IR/SliceSupport.cpp:82]:
```suggestion
/// Returns the predecessor branch operands that match `blockArg`, or 
/// nullopt if some of the predecessor terminators do not implement the
```
nit:

File: mlir/lib/IR/SliceSupport.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  if (block->isEntryBlock()) {
+    if (auto regionBranchOp =
+            dyn_cast<RegionBranchOpInterface>(block->getParentOp())) {
+      RegionSuccessor region(blockArg.getParentRegion());
+      SmallVector<Value> predecessorOperands = getRegionPredecessorOperands(
+          regionBranchOp, region, blockArg.getArgNumber());
+      return predecessorOperands;
+    }
+    // Unclear how to deal with this operation, conservatively return a
+    // failure.
```
Comment (gysit) [inline mlir/lib/IR/SliceSupport.cpp:134]:
```suggestion
    // If the interface is not implemented, there are no known control flow predecessors.
```
nit: I prefer a comment similar to the one above.

**Review by gysit — COMMENTED** [review]
Nice! 

I added some nit comments and can do a full review once moving the interface is done.

**Comment by Dinistro** [comment]
PR that moves the interface: https://github.com/llvm/llvm-project/pull/103927

File: mlir/include/mlir/Analysis/SliceWalk.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
+//===- SliceWalk.h - Helpers for performing IR slice walks ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_ANALYSIS_SLICEWALK_IR
+#define MLIR_ANALYSIS_SLICEWALK_IR
```
Comment (gysit) [inline mlir/include/mlir/Analysis/SliceWalk.h:10]:
```suggestion
#ifndef MLIR_ANALYSIS_SLICEWALK_H
#define MLIR_ANALYSIS_SLICEWALK_H
```
nit:

File: mlir/lib/Dialect/LLVMIR/Transforms/InlinerInterfaceImpl.cpp [hunk 9]
```diff
@@ -363,9 +323,14 @@ static void createNewAliasScopesFromNoAliasParameter(
 
       // Find the set of underlying pointers that this pointer is based on.
       SmallPtrSet<Value, 4> basedOnPointers;
-      for (Value pointer : pointerArgs)
-        llvm::copy(getUnderlyingObjectSet(pointer),
+      for (Value pointer : pointerArgs) {
+        FailureOr<SmallVector<Value>> underlyingObjectSet =
+            getUnderlyingObjectSet(pointer);
+        if (failed(underlyingObjectSet))
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/Transforms/InlinerInterfaceImpl.cpp:329]: Can you add a test case that exercise the failure case?

**Review by gysit — APPROVED** [review]
LGTM modulo nits!

File: mlir/include/mlir/Analysis/SliceWalk.h [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+/// Computes a vector of all control predecessors of `value`. Relies on
+/// RegionBranchOpInterface and BranchOpInterface to determine predecessors.
+/// Returns nullopt if `value` has no predecessors or when the relevant
+/// operations are missing the interface implementations.
+std::optional<SmallVector<Value>> getControlFlowPredecessors(Value value);
+
+} // namespace mlir
+
+#endif // MLIR_ANALYSIS_SLICEWALK_IR
```
Comment (gysit) [inline mlir/include/mlir/Analysis/SliceWalk.h:98]:
```suggestion
#endif // MLIR_ANALYSIS_SLICEWALK_H
```

**Pull request merged**